pkg/agent/core/wss.go

```diff
@@ -0,0 +1,93 @@
+package core
+
+// Working set size estimation
+// For more, see: https://www.notion.so/neondatabase/874ef1cc942a4e6592434dbe9e609350
+
+import (
+	"fmt"
+)
+
+type WssEstimatorConfig struct {
+	// MaxAllowedIncreaseFactor is the maximum tolerable increase in slope between windows.
+	// If the slope increases by more than this factor, we will cut off the working set size as the
+	// border between the two windows.
+	MaxAllowedIncreaseFactor float64
+	// InitialOffset is the index of the minimum working set size we must consider.
+	//
+	// In practice, this is taken from the scaling config's LFCMinWaitBeforeDownscaleMinutes, with
+	// the expectation that datapoints are all one minute apart, starting at 1m. So a value of 15m
+	// translates to an InitialOffset of 14 (-1 because indexes start at zero, but the first
+	// datapoint is 1m).
+	InitialOffset int
+	// WindowSize sets the number of datapoints averaged out in a slope calculation.
+	// This value must be >= 2.
+	//
+	// (Technically the datapoints aren't "averaged", but instead taken at an offset of
+	// ±WindowSize-1 -- although this is equivalent if the datapoints are monotonically
+	// non-decreasing.)
+	//
+	// In practice, this value is taken from the scaling config's LFCWindowSizeMinutes, with the
+	// expectation that datapoints are all one minute apart. So, a value of 5 minutes translates to
+	// a WindowSize of 5.
+	WindowSize int
+}
+
+// EstimateTrueWorkingSetSize returns an estimate of the "true" current working set size, given a
+// series of datapoints for the observed working set size over increasing time intervals.
+//
+// In practice, the 'series' is e.g., values of 'neon.lfc_approximate_working_set_size_seconds(d)'
+// for equidistant values of 'd' from 1 minute to 60 minutes.
+func EstimateTrueWorkingSetSize(
+	series []float64,
+	cfg WssEstimatorConfig,
+) float64 {
+	if cfg.WindowSize < 2 {
+		panic(fmt.Errorf("cfg.WindowSize must be >= 2 (got %v)", cfg.WindowSize))
+	}
+
+	w := cfg.WindowSize - 1
+
+	for t := cfg.InitialOffset; t < len(series)-w; t += 1 {
```

Comment:
If cfg.WindowSize > cfg.InitialOffset, is it possible to get out-of-bounds here?

Reply:
_Technically_ we wouldn't because if `t >= len(series)` then we fail the condition in the for loop, but we could _technically_ get out of bounds if `len(series) == 0`.

However, in `core/state.go` we already check to make sure we have enough datapoints:

https://github.com/neondatabase/autoscaling/blob/7daeb3ec97d7dc6849c3c7614c7906c91883f2e8/pkg/agent/core/state.go#L714-L716

Reply:
If InitialOffset=1, WindowSize=10, len(series)=100

then in the first iteration we will take `series[1] - series[1-9]` which is not good.

Reply:
Oh, good point, I guess I misread your comment :sweat_smile:

Added a check in 253c8f16239149426023a83be0c777b3c1746830